internal/fields/dependency_manager.go

```diff
@@ -0,0 +1,213 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package fields
+
+import (
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+	"gopkg.in/yaml.v3"
+
+	"github.com/elastic/elastic-package/internal/common"
+	"github.com/elastic/elastic-package/internal/configuration/locations"
+	"github.com/elastic/elastic-package/internal/logger"
+	"github.com/elastic/elastic-package/internal/packages/buildmanifest"
+)
+
+const (
+	ecsSchemaName      = "ecs"
+	gitReferencePrefix = "git@"
+
+	ecsSchemaFile = "fields.ecs.yml"
+	ecsSchemaURL  = "https://raw.githubusercontent.com/elastic/ecs/%s/generated/beats/%s"
+)
+
+// DependencyManager is responsible for resolving external field dependencies.
+type DependencyManager struct {
+	schema map[string][]FieldDefinition
+}
+
+// CreateFieldDependencyManager function creates a new instance of the DependencyManager.
+func CreateFieldDependencyManager(deps buildmanifest.Dependencies) (*DependencyManager, error) {
+	schema, err := buildFieldsSchema(deps)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't build fields schema")
+	}
+	return &DependencyManager{
+		schema: schema,
+	}, nil
+}
+
+func buildFieldsSchema(deps buildmanifest.Dependencies) (map[string][]FieldDefinition, error) {
+	schema := map[string][]FieldDefinition{}
+	ecsSchema, err := loadECSFieldsSchema(deps.ECS)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't load fields")
+	}
+	schema[ecsSchemaName] = ecsSchema
+	return schema, nil
+}
+
+func loadECSFieldsSchema(dep buildmanifest.ECSDependency) ([]FieldDefinition, error) {
+	if dep.Reference == "" {
+		logger.Debugf("ECS dependency isn't defined")
+		return nil, nil
+	}
+
+	gitReference, err := asGitReference(dep.Reference)
+	if err != nil {
+		return nil, errors.Wrap(err, "can't process the value as Git reference")
+	}
+
+	loc, err := locations.NewLocationManager()
+	if err != nil {
+		return nil, errors.Wrap(err, "error fetching profile path")
+	}
+
+	cachedSchemaPath := filepath.Join(loc.FieldsCacheDir(), ecsSchemaName, gitReference, ecsSchemaFile)
+	content, err := ioutil.ReadFile(cachedSchemaPath)
+	if err != nil && !errors.Is(err, os.ErrNotExist) {
+		return nil, errors.Wrapf(err, "can't read cached schema (path: %s)", cachedSchemaPath)
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		logger.Debugf("Pulling ECS dependency using reference: %s", dep.Reference)
+
+		url := fmt.Sprintf(ecsSchemaURL, gitReference, ecsSchemaFile)
+		logger.Debugf("Schema URL: %s", url)
+		resp, err := http.Get(url)
```

Comment:
:warning: here, is this the first network dependency we add to `elastic-package build`? I don't see it specially problematic, but it may be something we want to think twice.
It is probably better if we can avoid this, so developers workflow doesn't break if they are offline or if github is down.

Maybe an alternative is to have a set of subcommands to update the fields files, these commands would need connectivity, but the build process would continue being the same as now.

Reply:
Frankly speaking we already have network requirement with the stack command - Docker images for Elastic stack are mandatory. You can go offline if you have cached them already. The same policy is applied here. If the cache is populated with fields, you can perform offline builds.

>Maybe an alternative is to have a set of subcommands to update the fields files, these commands would need connectivity, but the build process would continue being the same ...

I would be cautious with such approach, as it complicates the process (with internet connectivity, it's just `build`). It also means that we have to store these imported definitions somewhere in the repository.

I'm happy to consider other possible approach if you have any specific workload in your mind :)

Reply:
Ok, let's go on like this, I don't see it specially problematic, we might revisit this in the future.

Not directly related to offline processes, but one thing that I like to think about a build process is that it generates artifacts without modifying the source files. In this case the line between source files and a built package is a bit blurry. We may need to think more about this in the frame of https://github.com/elastic/package-registry/issues/670 and https://github.com/elastic/package-registry/issues/168.